### PR TITLE
Added cmake build

### PR DESCRIPTION
**Pull request by yurp**
I've added ability to build the project with cmake. It's only support build now (doesn't support install)

### CODE REVIEW

File: lib/util/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+project(util)
+
+file(GLOB UTIL_SRC "fork_util.c"
+                   "fs_util.c"
+                   "io.c"
+                   "logger.c"
+                   "pseudo.c"
+                   "sys_util.c"
+                   "terminal.c")
```
Comment (iandinwoodie) [inline lib/util/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#dont-use-fileglob-in-projects): don't use glob in projects.

File: lib/tokenizer/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+FIX_OUTPUT_NAME_AND_ADD_FLEX_TARGET(go_lexer ${CMAKE_CURRENT_SOURCE_DIR}/golexer.l
+                                             ${CMAKE_CURRENT_BINARY_DIR}/golexer_fixed.l
+                                             ${CMAKE_CURRENT_BINARY_DIR}/golexer.c)
+
+include_directories(${CMAKE_CURRENT_BINARY_DIR}
+                    ${CMAKE_CURRENT_SOURCE_DIR}
+                    "../adt"
+                    "../util")
+
+file(GLOB TOKENIZER_SRC "tokenizer.c")
```
Comment (iandinwoodie) [inline lib/tokenizer/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#dont-use-fileglob-in-projects): don't use glob in projects.

File: lib/tgdb/tgdb-base/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+                   "tgdb_command.c"
+                   "tgdb_types.c")
+
+add_library(${PROJECT_NAME} STATIC ${TGDB_SRC})
+
+####################################################################################################
+# TESTS
+####################################################################################################
+
+file(GLOB TGDB_DRIVER_SRC "driver.c")
```
Comment (iandinwoodie) [inline lib/tgdb/tgdb-base/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#dont-use-fileglob-in-projects): don't use glob in projects.

File: lib/tgdb/tgdb-base/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+                    "../annotate-two"
+                    "../gdbmi"
+                    "../../adt"
+                    "../../rline"
+                    "../../util")
+
+file(GLOB TGDB_SRC "tgdb.c"
+                   "tgdb_client_interface.c"
+                   "tgdb_command.c"
+                   "tgdb_types.c")
```
Comment (iandinwoodie) [inline lib/tgdb/tgdb-base/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#dont-use-fileglob-in-projects): don't use glob in projects.

File: lib/tgdb/gdbmi/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,13 @@
+
+cmake_minimum_required(VERSION 2.8)
+
+project(tgdb_mi)
+
+include_directories("../tgdb-base"
+                    "../../adt"
+                    "../../util")
+
+file(GLOB MI_SRC "gdbmi_tgdb.c")
```
Comment (iandinwoodie) [inline lib/tgdb/gdbmi/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#dont-use-fileglob-in-projects): don't use glob in projects.

File: lib/tgdb/annotate-two/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+include_directories("../tgdb-base"
+                    "../../adt"
+                    "../../util")
+
+file(GLOB A2_SRC "a2-tgdb.c"
+                 "annotate.c"
+                 "commands.c"
+                 "data.c"
+                 "globals.c"
+                 "state_machine.c")
```
Comment (iandinwoodie) [inline lib/tgdb/annotate-two/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#dont-use-fileglob-in-projects): don't use glob in projects.

File: lib/rline/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,12 @@
+
+cmake_minimum_required(VERSION 2.8)
+
+project(rline)
+
+include_directories("../adt"
+                    "../util")
+
+file(GLOB RLINE_SRC "rline.c")
```
Comment (iandinwoodie) [inline lib/rline/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#dont-use-fileglob-in-projects): don't use glob in projects.

File: lib/kui/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+include_directories("../util"
+                    "../adt")
+
+add_library("kui" STATIC ${ADT_SRC})
+
+####################################################################################################
+# TESTS
+####################################################################################################
+
+file(GLOB KUI_DRIVER_SRC "kui_driver.c")
```
Comment (iandinwoodie) [inline lib/kui/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#dont-use-fileglob-in-projects): don't use glob in projects.

File: cgdb/interface.c [hunk 9]
```diff
@@ -1235,7 +1239,7 @@ int internal_if_input(int key)
                     /* Set the type of the command the user is typing in the status bar */
                     sbc_kind = SBC_NORMAL;
                     if_set_focus(CGDB_STATUS_BAR);
-                    /* Since the user is about to type in a command, allocate a buffer
```
Comment (iandinwoodie) [inline cgdb/interface.c]: Unnecessary whitespace changes.

File: lib/util/pseudo.c [hunk 10]
```diff
@@ -641,22 +644,22 @@ int pty_change_window_size(int masterfd, int row, int col, int xpixel,
...
+ * of the slave side of the pseudo terminal is stored in the buffer pointed
+ * to by slavename which must be able to hold at least 64 bytes.
  * slavenamesize is the size of the buffer pointed to by slavename. No
- * more than slavenamesize bytes will be written to slavename, including 
- * the terminating nul byte. If slave_termios is not null, it is passed to 
- * tcsetattr with the command TCSANOW to set the terminal attributes of the 
- * slave device. If slave_winsize is not null, it is passed to ioctl with 
- * the command TIOCSWINSZ to set the window size of the slave device. 
- * On success, returns 0 to the child process and returns the process 
- * id of the child process to the parent process. On error, returns -1 with
```
Comment (iandinwoodie) [inline lib/util/pseudo.c]: Unnecessary whitespace changes.

File: CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,74 @@
+
+cmake_minimum_required(VERSION 2.8)
```
Comment (iandinwoodie) [inline CMakeLists.txt]: If we were to adopt a build system, I'm not sure we would need to go with such an antiquated release (CMake 2.8 was released in 2009; 11 years ago).

File: CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+macro(FIX_OUTPUT_NAME_AND_ADD_FLEX_TARGET TARGET_NAME IN_NAME IN_FIXED_NAME OUT_NAME)
+    add_custom_command(
+        OUTPUT ${IN_FIXED_NAME}
+        COMMAND sed '/%option outfile="lex.yy.c"/d' ${IN_NAME} > ${IN_FIXED_NAME})
+    FLEX_TARGET(${TARGET_NAME} ${IN_FIXED_NAME} ${OUT_NAME})
+endmacro(FIX_OUTPUT_NAME_AND_ADD_FLEX_TARGET)
+
+####################################################################################################
+
+set(NCURSES_LIB "ncurses")
```
Comment (iandinwoodie) [inline CMakeLists.txt]: This doesn't seem to account the support for the `curses` lib.

File: CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+    -DPACKAGE_VERSION="20151129"
+    -DRETSIGTYPE=void
+    -DSETPGRP_VOID=1
+    -DSTDC_HEADERS=1
+    -DVERSION="20151129"
+    -DYYTEXT_POINTER=1)
+
+add_subdirectory(lib)
+add_subdirectory(cgdb)
+
```
Comment (iandinwoodie) [inline CMakeLists.txt]: Every top level CMake should include a project. Per [Modern CMake](https://cliutils.gitlab.io/modern-cmake/chapters/basics.html).

File: cgdb/CMakeLists.txt [hunk 14]
```diff
@@ -0,0 +1,50 @@
+
+cmake_minimum_required(VERSION 2.8)
+
+project(cgdb)
+
+# TODO: make configurable
```
Comment (iandinwoodie) [inline cgdb/CMakeLists.txt]: This is not an actionable `TODO` due to how generic (inexplicit) the note is.

File: cgdb/CMakeLists.txt [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+                   "filedlg.c"
+                   "filedlg.h"
+                   "highlight.c"
+                   "highlight_groups.c"
+                   "interface.c"
+                   "logo.c"
+                   "logo.h"
+                   "scroller.c"
+                   "sources.c"
+                   "usage.c")
```
Comment (iandinwoodie) [inline cgdb/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#dont-use-fileglob-in-projects): don't use glob in projects.

File: lib/adt/CMakeLists.txt [hunk 16]
```diff
@@ -0,0 +1,49 @@
+
+cmake_minimum_required(VERSION 2.8)
+
+file(GLOB ADT_SRC "ibuf.c"
+                  "queue.c"
+                  "tgdb_list.c"
+                  "std_bbtree.c"
+                  "std_btree.c"
+                  "std_hash.c"
+                  "std_list.c")
```
Comment (iandinwoodie) [inline lib/adt/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#dont-use-fileglob-in-projects): don't use glob in projects.

File: lib/adt/CMakeLists.txt [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+
+include_directories("../util")
+
+add_library("adt" STATIC ${ADT_SRC})
+
+####################################################################################################
+# TESTS
+####################################################################################################
+
+file(GLOB IBUF_DRIVER_SRC "ibuf_driver.c")
```
Comment (iandinwoodie) [inline lib/adt/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#dont-use-fileglob-in-projects): don't use glob in projects.

File: lib/adt/CMakeLists.txt [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+add_executable("ibuf_driver" ${IBUF_DRIVER_SRC})
+target_link_libraries("ibuf_driver" "adt" "util")
+
+####################################################################################################
+
+file(GLOB STD_HASH_DRIVER_SRC "std_hash_driver.c")
```
Comment (iandinwoodie) [inline lib/adt/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#dont-use-fileglob-in-projects): don't use glob in projects.

File: lib/adt/CMakeLists.txt [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+add_executable("std_hash_driver" ${STD_HASH_DRIVER_SRC})
+target_link_libraries("std_hash_driver" "adt" "util")
+
+####################################################################################################
+
+file(GLOB STD_LIST_DRIVER_SRC "std_list_driver.c")
```
Comment (iandinwoodie) [inline lib/adt/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#dont-use-fileglob-in-projects): don't use glob in projects.

File: lib/adt/CMakeLists.txt [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+add_executable("std_list_driver" ${STD_LIST_DRIVER_SRC})
+target_link_libraries("std_list_driver" "adt" "util")
+
+####################################################################################################
+
+file(GLOB STD_BTREE_DRIVER_SRC "std_btree_driver.c")
```
Comment (iandinwoodie) [inline lib/adt/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#dont-use-fileglob-in-projects): don't use glob in projects.

File: lib/adt/CMakeLists.txt [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+add_executable("std_btree_driver" ${STD_BTREE_DRIVER_SRC})
+target_link_libraries("std_btree_driver" "adt" "util")
+
+####################################################################################################
+
+file(GLOB STD_BBTREE_DRIVER_SRC "std_bbtree_driver.c")
```
Comment (iandinwoodie) [inline lib/adt/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#dont-use-fileglob-in-projects): don't use glob in projects.

File: lib/gdbmi/CMakeLists.txt [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+                                                ${CMAKE_CURRENT_BINARY_DIR}/gdbmi_lexer.c)
+
+ADD_FLEX_BISON_DEPENDENCY(gdbmi_lexer gdbmi_grammar)
+
+include_directories(${CMAKE_CURRENT_BINARY_DIR}
+                    ${CMAKE_CURRENT_SOURCE_DIR})
+
+file(GLOB GDBMI_SRC "gdbmi_oc.c"
+                    "gdbmi_pt.c"
+                    "gdbmi_parser.c")
```
Comment (iandinwoodie) [inline lib/gdbmi/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#dont-use-fileglob-in-projects): don't use glob in projects.

File: lib/gdbmi/CMakeLists.txt [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+
+add_library(gdbmi STATIC ${GDBMI_SRC}
+                         ${BISON_gdbmi_grammar_OUTPUTS}
+                         ${FLEX_gdbmi_lexer_OUTPUTS})
+
+####################################################################################################
+# TESTS
+####################################################################################################
+
+file(GLOB GDBMI_DRIVER_SRC "gdbmi_driver.c")
```
Comment (iandinwoodie) [inline lib/gdbmi/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#dont-use-fileglob-in-projects): don't use glob in projects.

File: lib/kui/CMakeLists.txt [hunk 24]
```diff
@@ -0,0 +1,22 @@
+
+cmake_minimum_required(VERSION 2.8)
+
+file(GLOB ADT_SRC "kui.c"
+                  "kui_term.c"
+                  "kui_tree.c")
```
Comment (iandinwoodie) [inline lib/kui/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#dont-use-fileglob-in-projects): don't use glob in projects.

File: cgdb/CMakeLists.txt [hunk 25]
```diff
@@ -0,0 +1,50 @@
...
+
+add_executable(${PROJECT_NAME} ${CGDB_SRC}
+                               ${FLEX_command_lexer_OUTPUTS})
```
Comment (iandinwoodie) [inline cgdb/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#avoid-custom-variables-in-the-arguments-of-project-commands): avoid custom variables in the arguments of project commands.

File: lib/adt/CMakeLists.txt [hunk 26]
```diff
@@ -0,0 +1,49 @@
...
+                  "queue.c"
+                  "tgdb_list.c"
+                  "std_bbtree.c"
+                  "std_btree.c"
+                  "std_hash.c"
+                  "std_list.c")
+
+include_directories("../util")
+
+add_library("adt" STATIC ${ADT_SRC})
```
Comment (iandinwoodie) [inline lib/adt/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#avoid-custom-variables-in-the-arguments-of-project-commands): avoid custom variables in the arguments of project commands.

File: lib/adt/CMakeLists.txt [hunk 27]
```diff
@@ -0,0 +1,49 @@
...
+
+####################################################################################################
+# TESTS
+####################################################################################################
+
+file(GLOB IBUF_DRIVER_SRC "ibuf_driver.c")
+add_executable("ibuf_driver" ${IBUF_DRIVER_SRC})
```
Comment (iandinwoodie) [inline lib/adt/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#avoid-custom-variables-in-the-arguments-of-project-commands): avoid custom variables in the arguments of project commands.

File: lib/adt/CMakeLists.txt [hunk 28]
```diff
@@ -0,0 +1,49 @@
...
+####################################################################################################
+
+file(GLOB STD_HASH_DRIVER_SRC "std_hash_driver.c")
+add_executable("std_hash_driver" ${STD_HASH_DRIVER_SRC})
+target_link_libraries("std_hash_driver" "adt" "util")
+
+####################################################################################################
+
+file(GLOB STD_LIST_DRIVER_SRC "std_list_driver.c")
+add_executable("std_list_driver" ${STD_LIST_DRIVER_SRC})
```
Comment (iandinwoodie) [inline lib/adt/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#avoid-custom-variables-in-the-arguments-of-project-commands): avoid custom variables in the arguments of project commands.

File: lib/adt/CMakeLists.txt [hunk 29]
```diff
@@ -0,0 +1,49 @@
...
+####################################################################################################
+
+file(GLOB STD_BTREE_DRIVER_SRC "std_btree_driver.c")
+add_executable("std_btree_driver" ${STD_BTREE_DRIVER_SRC})
+target_link_libraries("std_btree_driver" "adt" "util")
+
+####################################################################################################
+
+file(GLOB STD_BBTREE_DRIVER_SRC "std_bbtree_driver.c")
+add_executable("std_bbtree_driver" ${STD_BBTREE_DRIVER_SRC})
```
Comment (iandinwoodie) [inline lib/adt/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#avoid-custom-variables-in-the-arguments-of-project-commands): avoid custom variables in the arguments of project commands.

File: lib/gdbmi/CMakeLists.txt [hunk 30]
```diff
@@ -0,0 +1,37 @@
...
+include_directories(${CMAKE_CURRENT_BINARY_DIR}
+                    ${CMAKE_CURRENT_SOURCE_DIR})
+
+file(GLOB GDBMI_SRC "gdbmi_oc.c"
+                    "gdbmi_pt.c"
+                    "gdbmi_parser.c")
+
+add_library(gdbmi STATIC ${GDBMI_SRC}
+                         ${BISON_gdbmi_grammar_OUTPUTS}
+                         ${FLEX_gdbmi_lexer_OUTPUTS})
```
Comment (iandinwoodie) [inline lib/gdbmi/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#avoid-custom-variables-in-the-arguments-of-project-commands): avoid custom variables in the arguments of project commands.

File: lib/gdbmi/CMakeLists.txt [hunk 31]
```diff
@@ -0,0 +1,37 @@
...
+
+####################################################################################################
+# TESTS
+####################################################################################################
+
+file(GLOB GDBMI_DRIVER_SRC "gdbmi_driver.c")
+add_executable("gdbmi_driver" ${GDBMI_DRIVER_SRC})
```
Comment (iandinwoodie) [inline lib/gdbmi/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#avoid-custom-variables-in-the-arguments-of-project-commands): avoid custom variables in the arguments of project commands.

File: lib/kui/CMakeLists.txt [hunk 32]
```diff
@@ -0,0 +1,22 @@
...
+
+include_directories("../util"
+                    "../adt")
+
+add_library("kui" STATIC ${ADT_SRC})
```
Comment (iandinwoodie) [inline lib/kui/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#avoid-custom-variables-in-the-arguments-of-project-commands): avoid custom variables in the arguments of project commands.

File: lib/kui/CMakeLists.txt [hunk 33]
```diff
@@ -0,0 +1,22 @@
...
+
+####################################################################################################
+# TESTS
+####################################################################################################
+
+file(GLOB KUI_DRIVER_SRC "kui_driver.c")
+add_executable("kui_driver" ${KUI_DRIVER_SRC})
```
Comment (iandinwoodie) [inline lib/kui/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#avoid-custom-variables-in-the-arguments-of-project-commands): avoid custom variables in the arguments of project commands.

File: lib/rline/CMakeLists.txt [hunk 34]
```diff
@@ -0,0 +1,12 @@
...
+
+add_library(${PROJECT_NAME} STATIC ${RLINE_SRC})
```
Comment (iandinwoodie) [inline lib/rline/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#avoid-custom-variables-in-the-arguments-of-project-commands): avoid custom variables in the arguments of project commands.

File: lib/tgdb/annotate-two/CMakeLists.txt [hunk 35]
```diff
@@ -0,0 +1,18 @@
...
+
+add_library(${PROJECT_NAME} STATIC ${A2_SRC})
```
Comment (iandinwoodie) [inline lib/tgdb/annotate-two/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#avoid-custom-variables-in-the-arguments-of-project-commands): avoid custom variables in the arguments of project commands.

File: lib/tgdb/gdbmi/CMakeLists.txt [hunk 36]
```diff
@@ -0,0 +1,13 @@
...
+
+add_library(${PROJECT_NAME} STATIC ${MI_SRC})
```
Comment (iandinwoodie) [inline lib/tgdb/gdbmi/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#avoid-custom-variables-in-the-arguments-of-project-commands): avoid custom variables in the arguments of project commands.

File: lib/tgdb/tgdb-base/CMakeLists.txt [hunk 37]
```diff
@@ -0,0 +1,37 @@
...
+
+add_library(${PROJECT_NAME} STATIC ${TGDB_SRC})
```
Comment (iandinwoodie) [inline lib/tgdb/tgdb-base/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#avoid-custom-variables-in-the-arguments-of-project-commands): avoid custom variables in the arguments of project commands.

File: lib/tgdb/tgdb-base/CMakeLists.txt [hunk 38]
```diff
@@ -0,0 +1,37 @@
...
+
+####################################################################################################
+# TESTS
+####################################################################################################
+
+file(GLOB TGDB_DRIVER_SRC "driver.c")
+add_executable("driver" ${TGDB_DRIVER_SRC})
```
Comment (iandinwoodie) [inline lib/tgdb/tgdb-base/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#avoid-custom-variables-in-the-arguments-of-project-commands): avoid custom variables in the arguments of project commands.

File: lib/tokenizer/CMakeLists.txt [hunk 39]
```diff
@@ -0,0 +1,39 @@
...
+
+add_library(tokenizer STATIC ${TOKENIZER_SRC}
+                             ${FLEX_ada_lexer_OUTPUTS}
+                             ${FLEX_c_lexer_OUTPUTS}
+                             ${FLEX_d_lexer_OUTPUTS}
+                             ${FLEX_go_lexer_OUTPUTS})
```
Comment (iandinwoodie) [inline lib/tokenizer/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#avoid-custom-variables-in-the-arguments-of-project-commands): avoid custom variables in the arguments of project commands.

File: lib/tokenizer/CMakeLists.txt [hunk 40]
```diff
@@ -0,0 +1,39 @@
...
+
+####################################################################################################
+# TESTS
+####################################################################################################
+
+file(GLOB TOKENIZER_DRIVER_SRC "tokenizer_driver.c")
+add_executable("tokenizer_driver" ${TOKENIZER_DRIVER_SRC})
```
Comment (iandinwoodie) [inline lib/tokenizer/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#avoid-custom-variables-in-the-arguments-of-project-commands): avoid custom variables in the arguments of project commands.

File: lib/util/CMakeLists.txt [hunk 41]
```diff
@@ -0,0 +1,15 @@
...
+
+add_library(${PROJECT_NAME} STATIC ${UTIL_SRC})
```
Comment (iandinwoodie) [inline lib/util/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#avoid-custom-variables-in-the-arguments-of-project-commands): avoid custom variables in the arguments of project commands.

File: lib/tokenizer/CMakeLists.txt [hunk 42]
```diff
@@ -0,0 +1,39 @@
...
+                             ${FLEX_ada_lexer_OUTPUTS}
+                             ${FLEX_c_lexer_OUTPUTS}
+                             ${FLEX_d_lexer_OUTPUTS}
+                             ${FLEX_go_lexer_OUTPUTS})
+
+####################################################################################################
+# TESTS
+####################################################################################################
+
+file(GLOB TOKENIZER_DRIVER_SRC "tokenizer_driver.c")
```
Comment (iandinwoodie) [inline lib/tokenizer/CMakeLists.txt]: Per [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#dont-use-fileglob-in-projects): don't use glob in projects.

**Comment by iandinwoodie** [comment]
@yurp I do recognize that, since this PR has been sitting here for so long, the "Modern CMake" documents I refer to in the review may not have been around when this was originally opened. However, I felt it was best to perform my review with comparison to current practices.

**Comment by KOLANICH** [comment]
IMHO
1. Use of GLOB is OK. It saves us from editing CMakeLists.txt too often.
2. Use of variables is completely OK. And inevitable when `GLOB` is used.
3. CMake is a build system which one of goals is portability and compatibility. It has some batteries making compatibility to different compilers included. It is the main reason to use it.

**Comment by iandinwoodie** [comment]
> IMHO
> 
> 1. Use of GLOB is OK. It saves us from editing CMakeLists.txt too often.
> 2. Use of variables is completely OK. And inevitable when `GLOB` is used.

Yes, points 1 and 2 here go hand-in-hand, so we can talk about them as a single topic. Of course `GLOB` saves people from editing `CMakeLists.txt` too often; its a convenience feature. However, the feature has problematic downsides. Here is the advice from the CMake developers themselves:

>**Note:** *We do not recommend using GLOB to collect a list of source files from your source tree. If no CMakeLists.txt file changes when a source is added or removed then the generated build system cannot know when to ask CMake to regenerate. The CONFIGURE_DEPENDS flag may not work reliably on all generators, or if a new generator is added in the future that cannot support it, projects using it will be stuck. Even if CONFIGURE_DEPENDS works reliably, there is still a cost to perform the check on every rebuild.*

([reference](https://cmake.org/cmake/help/git-master/command/file.html#filesystem))

So, you may choose to use the `GLOB` feature, but it is a practice that is discouraged by the tool maintainers themselves. As a result, I still believe my claim that "the proposed CMake files do not adhere to the best practices of modern CMake" holds true.

> 3. CMake is a build system which one of goals is portability and compatibility. It has some batteries making compatibility to different compilers and distros included. It is the main reason to use it.

CMake is a cross-platform build system and it does that very well. However, there is a difference between being cross-platform and portable.

My original comment was: "the author provides no guarantee of **portability**." I also included a specific example in my original comment as well: "The proposed changes seem to suit only submitter's environment (e.g., there was no sign of support for curses or ncursesw with wide character support)."

To elaborate: I am not against a migration to CMake; it would vastly simply the build system which would reduce work for maintainers. However, for such a change to happen we would need some sort of guarantee that there is not a loss in **portability**. From the review here I pointed out two scenarios (`libcurses` or `libncursesw`) where portability would have been broken. All current non-default autoconf configurations and tests would need to be ported over to some CMake equivalent.

I hope this clarified the comments made in my review.

**Comment by iandinwoodie** [comment]
@KOLANICH , I have posted [a response](https://github.com/cgdb/cgdb/pull/77#issuecomment-584445520) to [your comment](https://github.com/cgdb/cgdb/pull/77#issuecomment-583989785) in this thread.

**Comment by KOLANICH** [comment]
>We do not recommend using GLOB to collect a list of source files from your source tree. If no CMakeLists.txt file changes when a source is added or removed then the generated build system cannot know when to ask CMake to regenerate.

IMHO it's not a very big issue, we are usually calling cmake explicitly, so it always regenerates.

>As a result, I still believe my claim that "the proposed CMake files do not adhere to the best practices of modern CMake" holds true.

Yes, but it doesn't mean the practices described in the book are really the best. What is the best is decided on per-project basis.

>However, for such a change to happen we would need some sort of guarantee that there is not a loss in portability. From the review here I pointed out two scenarios (libcurses or libncursesw) where portability would have been broken.

I have looked the source code, the author is really doing lots of things wrong. The most obvious changes:
1. he is using `GLOB` wrong, because he enumerates file names instead of providing their templates
2. he should really have used https://cmake.org/cmake/help/latest/module/FindCurses.html . It is one of the best things about CMake that it has lots of modules for discovering polular open-souece libs in its standard library.

**Comment by iandinwoodie** [comment]
Closing due to:

- Unsigned commits.
- Issues identified by user @KOLANICH and myself.
